### PR TITLE
Move code from qml.utils.sparse_hamiltonian to Hamiltonian.sparse_matrix #3236

### PR DESCRIPTION
**Pull request by 12mB7693**
…trix, added a deprecation warning to qml.utils.sparse_hamiltonian

### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/ops/qubit/hamiltonian.py [hunk 1]
```diff
@@ -20,12 +20,15 @@
 import numbers
 from collections.abc import Iterable
 from copy import copy
+import scipy
+import functools
 from typing import List
```
Comment (albi3ro) [inline pennylane/ops/qubit/hamiltonian.py:25]:
```suggestion
import functools
from typing import List


import scipy


```

File: pennylane/ops/qubit/hamiltonian.py [hunk 2]
```diff
@@ -342,6 +345,87 @@ def compute_grouping(self, grouping_type="qwc", method="rlf"):
...
+
+            temp_mats.append(red_mat.tocsr())
+            # Value of 100 arrived at empirically to balance time savings vs memory use. At this point
+            # the `temp_mats` are summed into the final result and the temporary storage array is
+            # cleared.
+            if (len(temp_mats) % 100) == 0:
+                matrix += sum(temp_mats)
+                temp_mats = []
+
+        matrix += sum(temp_mats)
```
Comment (albi3ro) [inline pennylane/ops/qubit/hamiltonian.py:426]:
```suggestion
        matrix += sum(temp_mats)
```

File: pennylane/utils.py [hunk 3]
```diff
@@ -20,6 +20,9 @@
 import functools
 import inspect
 import numbers
+from operator import matmul
```
Comment (albi3ro) [inline pennylane/utils.py:23]:
```suggestion
```

**Review by albi3ro — CHANGES_REQUESTED** [review]
Thanks for opening this PR @12mB7693 !

The tests for the original `sparse_hamiltonian` method in [`test_utils.py`](https://github.com/PennyLaneAI/pennylane/blob/880517228462ba01af35d94204e55fc9c47a7cac/tests/test_utils.py#L67) can be copied and repurposed to be added to [`tests/ops/qubit/test_hamiltonian.py`](https://github.com/PennyLaneAI/pennylane/blob/master/tests/ops/qubit/test_hamiltonian.py).

We also need a changelog entry under "Deprecations". Don't forget to add your name to the list of contributors as well! The changelog can be found at: https://github.com/PennyLaneAI/pennylane/blob/master/doc/releases/changelog-dev.md

We will also need to add add entry to the [`doc/developement/deprecations.rst`](https://github.com/PennyLaneAI/pennylane/blob/880517228462ba01af35d94204e55fc9c47a7cac/doc/development/deprecations.rst) file. Since are now developing for v0.29, let's schedule `qml.utils.sparse_hamiltonian` for removal in v0.31.

I've left comments with suggestions to fix the various linting and formatting comments.  If you want to learn more about running the various tools we use to standardize code formatting, you can read [our pull request guide](https://docs.pennylane.ai/en/stable/development/guide/pullrequests.html).


We can also do this is a later PR if necessary, but we can now update the three places in PennyLane that use `qml.utils.sparse_hamiltonian`:
* https://github.com/PennyLaneAI/pennylane/blob/880517228462ba01af35d94204e55fc9c47a7cac/pennylane/devices/default_qubit.py#L590
* https://github.com/PennyLaneAI/pennylane/blob/880517228462ba01af35d94204e55fc9c47a7cac/pennylane/ops/functions/matrix.py#L125
* https://github.com/PennyLaneAI/pennylane/blob/880517228462ba01af35d94204e55fc9c47a7cac/pennylane/optimize/lie_algebra.py#L62

File: pennylane/devices/default_qubit.py [hunk 4]
```diff
@@ -587,7 +587,7 @@ def expval(self, observable, shot_range=None, bin_size=None):
                 # Coefficients and the state are not trainable, we can be more
                 # efficient in how we compute the Hamiltonian sparse matrix.
                 if observable.name == "Hamiltonian":
-                    Hmat = qml.utils.sparse_hamiltonian(observable, wires=self.wires)
+                    Hmat = observable.sparse_matrix(wire_order=self.wires)
                 elif observable.name == "SparseHamiltonian":
                     Hmat = observable.sparse_matrix()
```
Comment (albi3ro) [inline pennylane/devices/default_qubit.py:592]:
```suggestion
                if observable.name in {"Hamiltonian", "SparseHamiltonian"}:
                     Hmat = observable.sparse_matrix()
```
An optional simplification we can now make.

**Review by albi3ro — COMMENTED** [review]
Thanks for the updates. This looks in really good shape.

I'd just like to make a comment about handling the wire order at the end with `expand_matrix`, versus constructing the sparse matrix with the wire order from the start.

Even for a simple case like:
```
H = 2.0 * qml.PauliX(0) @ qml.PauliY(1) @ qml.PauliZ(2) @ qml.PauliX(3)
%timeit qml.utils.sparse_hamiltonian(H, wires = list(reversed(H.wires)))
```

You'll see that using that handling the wire order at the end is quite a bit slower. We want to construct to the specific wire order from the very beginning.

So instead of using `self.wires` in`Hamiltonian.sparse_matrix`, we want to use `wires = wire_order or self.wires`.

Does that make sense?

**Comment by 12mB7693** [comment]
@albi3ro Thank you for reviewing what I did and for your comments, that helped me a lot.
I think I get what you mean and I will push an update soon. 
Shall I actually also update the tests, where the old function qml.utils.sparse_hamiltonian(..) is called?

**Comment by albi3ro** [comment]
> @albi3ro Thank you for reviewing what I did and for your comments, that helped me a lot.
> I think I get what you mean and I will push an update soon.
> Shall I actually also update the tests, where the old function qml.utils.sparse_hamiltonian(..) is called?

Updating the tests where `qml.utils.spase_hamiltonian` is called (outside of unit tests) would be great! But I won't necessarily block approval based on that being accomplished.

**Comment by 12mB7693** [comment]
> > @albi3ro Thank you for reviewing what I did and for your comments, that helped me a lot.
> > I think I get what you mean and I will push an update soon.
> > Shall I actually also update the tests, where the old function qml.utils.sparse_hamiltonian(..) is called?
> 
> Updating the tests where `qml.utils.spase_hamiltonian` is called (outside of unit tests) would be great! But I won't necessarily block approval based on that being accomplished.

I'm not sure if I understood you correctly regarding the updates of the tests. Do you mean with "outside of unit tests" that I should not update the unit tests?

File: pennylane/devices/default_qubit.py [hunk 5]
```diff
@@ -586,9 +586,7 @@ def expval(self, observable, shot_range=None, bin_size=None):
             else:
                 # Coefficients and the state are not trainable, we can be more
                 # efficient in how we compute the Hamiltonian sparse matrix.
-                if observable.name == "Hamiltonian":
-                    Hmat = qml.utils.sparse_hamiltonian(observable, wires=self.wires)
-                elif observable.name == "SparseHamiltonian":
+                if observable.name in {"Hamiltonian", "SparseHamiltonian"}:
                     Hmat = observable.sparse_matrix()
```
Comment (albi3ro) [inline pennylane/devices/default_qubit.py:590]: Sorry this is my mistake in what I suggested:
```suggestion
                    Hmat = observable.sparse_matrix(wires=self.wires)
```

This should fix some of the failing tests.

**Comment by 12mB7693** [comment]
Not sure why the documentation check fails, I ran "make doc" locally and I get the warning
`releases/changelog-dev.md:391: WARNING: Definition list ends without a blank line; unexpected unindent.
`
However, I do not see how to fix it.

File: doc/releases/changelog-dev.md [hunk 6]
```diff
@@ -363,6 +368,10 @@
 
 <h3>Deprecations</h3>
 
+* `qml.utils.sparse_hamiltonian` function has been deprecated, and usage will now raise a warning.
+   Instead, one should use the `~.Hamiltonian.sparse_matrix` method.
+  [(#3236)](https://github.com/PennyLaneAI/pennylane/issues/3236)
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:373]:
```suggestion
* `qml.utils.sparse_hamiltonian` function has been deprecated, and usage will now raise a warning.
  Instead, one should use the `~.Hamiltonian.sparse_matrix` method.
  [(#3236)](https://github.com/PennyLaneAI/pennylane/issues/3236)
```

Maybe the space was the problem?

File: doc/releases/changelog-dev.md [hunk 7]
```diff
@@ -287,6 +287,11 @@
   `qml.gradients.SUPPORTED_GRADIENT_KWARGS`.
   [(#3526)](https://github.com/PennyLaneAI/pennylane/pull/3526)
 
+* Moved `qml.utils.sparse_hamiltonian` function to `~.Hamiltonian.sparse_matrix` method.
+  [(#3236)](https://github.com/PennyLaneAI/pennylane/issues/3236)
+
+* Improve the `PauliSentence.operation()` method to avoid instantiating an `SProd` operator when
+
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:294]:
```suggestion
```

File: doc/releases/changelog-dev.md [hunk 8]
```diff
@@ -287,6 +287,9 @@
   `qml.gradients.SUPPORTED_GRADIENT_KWARGS`.
   [(#3526)](https://github.com/PennyLaneAI/pennylane/pull/3526)
 
+* Moved `qml.utils.sparse_hamiltonian` function to `~.Hamiltonian.sparse_matrix` method.
+  [(#3236)](https://github.com/PennyLaneAI/pennylane/issues/3236)
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:291]:
```suggestion
  [(#3236)](https://github.com/PennyLaneAI/pennylane/pull/3585)
```

File: doc/releases/changelog-dev.md [hunk 9]
```diff
@@ -363,6 +366,10 @@
 
 <h3>Deprecations</h3>
 
+* `qml.utils.sparse_hamiltonian` function has been deprecated, and usage will now raise a warning.
+  Instead, one should use the `~.Hamiltonian.sparse_matrix` method.
+  [(#3236)](https://github.com/PennyLaneAI/pennylane/issues/3236)
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:371]:
```suggestion
  [(#3585)](https://github.com/PennyLaneAI/pennylane/pull/3585/)
```

File: doc/releases/changelog-dev.md [hunk 10]
```diff
@@ -287,6 +287,9 @@
   `qml.gradients.SUPPORTED_GRADIENT_KWARGS`.
   [(#3526)](https://github.com/PennyLaneAI/pennylane/pull/3526)
 
+* Moved `qml.utils.sparse_hamiltonian` function to `~.Hamiltonian.sparse_matrix` method.
+  [(#3585 )](https://github.com/PennyLaneAI/pennylane/pull/3585)
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:291]:
```suggestion
  [(#3585)](https://github.com/PennyLaneAI/pennylane/pull/3585)
```

File: doc/releases/changelog-dev.md [hunk 11]
```diff
@@ -412,6 +415,10 @@
 
 <h3>Deprecations</h3>
 
+* `qml.utils.sparse_hamiltonian` function has been deprecated, and usage will now raise a warning.
+  Instead, one should use the `~.Hamiltonian.sparse_matrix` method.
+  [(#3585)](https://github.com/PennyLaneAI/pennylane/pull/3585/)
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:420]:
```suggestion
  [(#3585)](https://github.com/PennyLaneAI/pennylane/pull/3585)
```

WHY SPHINX WHY

**Review by albi3ro — APPROVED** [review]
Thanks @12mB7693 ! 🎉 

Looks good to me.  Glad to finally have this tidied up :)

Sphinx may still be causing some problems, but I'll try and fix then if I can.

Excellent work :)

**Comment by 12mB7693** [comment]
No problem, I'm glad that I could help. This task was definitely a good starting point as a new contributor. 
Thank you for your support @albi3ro!

**Pull request merged**